sympy/matrices/random.py

```diff
@@ -0,0 +1,2378 @@
+        raise ValueError(
+            "index argument must be tuple of two matrix index integer.")
+
+    if row == col:
+        # identity or _multiply by scalar
+        obj[row, col] = value or 1
+        if scalars:
+            raise ValueError(
+                "if row==col no further args than value may be supplied.")
+    elif not scalars:
+        # elementary
+        if value is None:
+            # transposition
+            obj[row, row] = 0
+            obj[row, col] = 1
+            obj[col, row] = 1
+            obj[col, col] = 0
+        else:
+            # add scalar multiple to another
+            obj[row, col] = value
+    elif len(scalars) == 3:
+        obj[row, row] = value
+        y, v, u = scalars
+        obj[row, col] = y
+        obj[col, row] = -v
+        obj[col, col] = u
+    else:
+        msg = "either no, one or three additional scalars arguments " \
+              "may be supplied, but not %i"
+        raise ValueError(msg % len(scalars))
+    return obj
+
+
+# === fundamental matrix functions ===
+
+
+def complex_to_real(mat=None):
+    r""" returns real version 2n x 2n of a complex matrix n x n
+
+    Explanation
+    ===========
+    Any complex vector space $V$ with basis $(b_1, \dots, b_n)$
+    can be seen as a real vector space $V^{\prime}$ with basis
+    $(b_1, b_1 \ i, \dots, b_n, b_n \ i)$
+    of dimension $2 \ n$.
+
+    Since multiplication with complex number is linear on $V$
+    resp. $V^{\prime}$,
+    any complex homomorphism
+    (i.e. complex matrix $\mathbf{A}$) on $V$
+    becomes a real homomorphism
+    (i.e. matrix $\mathbf{A}^{\prime}$) on $V^{\prime}$.
+
+    This transforms a complex matrix $\mathbf{A}$
+    to the corresponding real matrix $\mathbf{A}^{\prime}$.
+
+    Examples
+    ========
+    >>> from sympy import Matrix, I
+    >>> from sympy.abc import a, b
+    >>> from sympy.matrices.random import complex_to_real
+
+    >>> z = a + b * I
+    >>> A = Matrix([[z]])
+    >>> A
+    Matrix([[a + I*b]])
+
+    >>> complex_to_real(A)
+    Matrix([
+    [ re(a) - im(b), re(b) + im(a)],
+    [-re(b) - im(a), re(a) - im(b)]])
+
+    >>> A = Matrix([[z, 0],[1, I]])
+    >>> A
+    Matrix([
+    [a + I*b, 0],
+    [      1, I]])
+    >>> complex_to_real(A)
+    Matrix([
+    [ re(a) - im(b), re(b) + im(a),  0, 0],
+    [-re(b) - im(a), re(a) - im(b),  0, 0],
+    [             1,             0,  0, 1],
+    [             0,             1, -1, 0]])
+
+    Parameters
+    ==========
+    mat : Matrix
+        a complex matrix
+
+    """
+    dim = _shape(mat)[0]
+    mat = mat or _eye(dim)
+    obj = super_elementary_matrix(2 * dim)
+    for i in range(dim):
+        for j in range(dim):
+            z_value = mat[i, j]
+            a, b = _re(z_value), _im(z_value)
+            obj[2 * i, 2 * j] = a
+            obj[2 * i, 2 * j + 1] = b
+            obj[2 * i + 1, 2 * j] = -b
+            obj[2 * i + 1, 2 * j + 1] = a
+    return obj
```

Comment:
It should be possible to do this with something like:
```python
M = Matrix([[a + I*b, 0], [1, I]])

Mr, Mi = M.as_real_imag()

Mplus = Mr + Mi
Mminus = Mr - Mi
Mnew = zeros(2*Mr.rows, 2*Mr.cols)
Mnew[::2,::2] = Mminus
Mnew[1::2,1::2] = Mminus
Mnew[1::2,::2] = -Mplus
Mnew[::2,1::2] = Mplus
```
It is preferable to use higher level operations like addition, slicing etc because they can be optimised for sparse matrices.

The code shown above doesn't work because of a bug though.

Reply:
> The code shown above doesn't work because of a bug though.

I mean that the code that I wrote does not work because of a bug:
```python
In [51]: M = zeros(4, 4)

In [52]: M[::2,::2] = ones(2, 2)
---------------------------------------------------------------------------
ShapeError 
```

Reply:
Slices seem not to work but I came up with a different approach permuting block matrices
```
    a, b = mat.as_real_imag()
    M = zeros(2 * dim)
    M[0, 0] = a
    M[0, dim] = b
    M[dim, 0] = -b
    M[dim, dim] = a

    p = list()
    for i in range(dim):
        p.extend([i, dim + i])
    M = M.permute(p).T.permute(p).T
```

Reply:
The slicing assignment should be fixed. I'll open a PR for that.